tests/test-update-direct.c

```diff
@@ -0,0 +1,255 @@
+  g_autoptr(GError) error = NULL;
+
+  switch (state)
+    {
+      case EOS_UPDATER_STATE_NONE:
+      case EOS_UPDATER_STATE_ERROR:
+      case EOS_UPDATER_STATE_READY:
+      case EOS_UPDATER_STATE_UPDATE_AVAILABLE:
+      case EOS_UPDATER_STATE_UPDATE_READY:
+      case EOS_UPDATER_STATE_UPDATE_APPLIED:
+        should_succeed = FALSE;
+        break;
+      case EOS_UPDATER_STATE_POLLING:
+      case EOS_UPDATER_STATE_FETCHING:
+      case EOS_UPDATER_STATE_APPLYING_UPDATE:
+      case EOS_UPDATER_N_STATES: /* just supress the warning about covering all enum values... */
+      default:
+        should_succeed = TRUE;
+        break;
+    }
+
+  g_debug ("Trying to cancel state %s", states_str[state]);
+
+  cancelled_states[state] = TRUE;
+  eos_updater_call_cancel_sync (updater, NULL, &error);
+
+  if (should_succeed)
+    {
+      g_assert_no_error (error);
+      ++cancel_calls_count;
+      g_debug ("Cancelled state %s", states_str[state]);
+    }
+  else
+    {
+      g_assert (error != NULL);
+      g_debug ("Error cancelling %s: %s", states_str[state], error->message);
+    }
+
+  return should_succeed;
+}
+
+static void
+updater_state_changed_cb (EosUpdater *updater,
+                          GParamSpec *pspec,
+                          gpointer data)
+{
+  EosUpdaterState state = eos_updater_get_state (updater);
+  GMainLoop *loop = (GMainLoop *) data;
+
+  /* we call the Cancel() method from the EOS Updater on every
+   * state once (it will either perform the cancel or return an error
+   * depending on the state); when a cancel has been called on a state already
+   * (or it gets an error), we call the next step in the update logic */
+  if (!cancelled_states[state] && cancel_update (updater))
+    return;
+
+  g_debug ("State changed %s", states_str[state]);
+  switch (state)
+    {
+      case EOS_UPDATER_STATE_ERROR:
+        if (g_strcmp0 (eos_updater_get_error_name (updater),
+                       "com.endlessm.Updater.Error.Cancelled") == 0)
+          ++cancelled_error_count;
+      case EOS_UPDATER_STATE_NONE:
+      case EOS_UPDATER_STATE_READY:
+        eos_updater_call_poll (updater, NULL, NULL, NULL);
+        break;
+      case EOS_UPDATER_STATE_UPDATE_AVAILABLE:
+        eos_updater_call_fetch (updater, NULL, NULL, NULL);
+        break;
+      case EOS_UPDATER_STATE_UPDATE_READY:
+        eos_updater_call_apply (updater, NULL, NULL, NULL);
+        break;
+      case EOS_UPDATER_STATE_UPDATE_APPLIED:
+        g_main_loop_quit (loop);
+        break;
+      case EOS_UPDATER_STATE_POLLING:
+      case EOS_UPDATER_STATE_FETCHING:
+      case EOS_UPDATER_STATE_APPLYING_UPDATE:
+      case EOS_UPDATER_N_STATES: /* just supress the warning about covering all enum values... */
+      default:
+        /* let it go until the next state change occurs */
+        break;
+    }
+}
+
+static void
+test_cancel_update (EosUpdaterFixture *fixture,
+                    gconstpointer user_data)
```

Comment:
For each unit test (so, just the one in this file), please add a comment directly above the test function which briefly explains what the test checks. (I realise other tests in eos-updater probably don’t do this, but it’s something I’m trying to introduce).

For example:
```
/* Test that calling Cancel() on each state in a complete update correctly causes a transition to the Error state. */
static void
test_cancel_update (…)
```